---
title: "Proteomics Data in Annotation Hub"
output:
  BiocStyle::html_document:
     toc: false
     toc_depth: 1
vignette: >
  % \VignetteIndexEntry{Proteomics Data in Annotation Hub}
  % \VignetteEngine{knitr::rmarkdown}
  % \VignetteKeyword{proteomics, mass spectrometry, data}
  % \VignettePackage{ProteomicsAnnotationHubData}
---


```{r style, echo = FALSE, results = 'asis', message=FALSE}
BiocStyle::markdown()
```

**Package**: `r Biocpkg("ProteomicsAnnotationHubData")`<br />
**Authors**: `r packageDescription("ProteomicsAnnotationHubData")[["Author"]] `<br />
**Modified:** `r file.info("ProteomicsAnnotationHubData.Rmd")$mtime`<br />
**Compiled**: `r date()`


# Introduction

About `r Biocpkg("AnnotationHub")`:

> This package provides a client for the Bioconductor AnnotationHub
> web resource. The AnnotationHub web resource provides a central
> location where genomic files (e.g., VCF, bed, wig) and other
> resources from standard locations (e.g., UCSC, Ensembl) can be
> discovered. The resource includes metadata about each resource,
> e.g., a textual description, tags, and date of modification. The
> client creates and manages a local cache of files retrieved by the
> user, helping with quick and reproducible access.

The goal of this project is to expand this functionality to mass
spectrometry and proteomics.


See the `AnnotationHub`'s
[How-To's](http://bioconductor.org/packages/devel/bioc/vignettes/AnnotationHub/inst/doc/AnnotationHub-HOWTO.html)
and
[Access the AnnotationHub Web Service](http://bioconductor.org/packages/devel/bioc/vignettes/AnnotationHub/inst/doc/AnnotationHub.html)
vignettes for a description on how to use it.

**TODO**: add example code for proteomics once the first data is available.

# Available datasets

## PXD000001

Description

> Expected reporter ion ratios: Erwinia peptides: 1:1:1:1:1:1 Enolase
> spike (sp|P00924|ENO1_YEAST): 10:5:2.5:1:2.5:10 BSA spike
> (sp|P02769|ALBU_BOVIN): 1:2.5:5:10:5:1 PhosB spike
> (sp|P00489|PYGM_RABIT): 2:2:2:2:1:1 Cytochrome C spike
> (sp|P62894|CYC_BOVIN): 1:1:1:1:1:2.

Four data files from the PRIDE
[PXD000001](http://www.ebi.ac.uk/pride/archive/projects/PXD000001)
experiment are served through `AnnotationHub`.

1. The raw mass spectrometry data, via the
   `TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML`
   file from the PRDIE ftp site, served as an `mzRpwiz` object, from
   the `r Biocpkg("mzR")` package.

2. The `mzTab` file, via the `F063721.dat-mztab.txt` file from the
   PRIDE ftp site, served as an `MSnSet` object, from the
   `r Biocpkg("MSnbase")` package.

3. The protein data base, via the `erwinia_carotovora.fasta` file from
	the PRIDE ftp server, served as a `AAStringSet` object, from the
	`r Biocpkg("Biostrings")` package.

4. The identification results, produced using the `MSGF+` search
   engine, served as a `mzRident` object, from the `r Biocpkg("mzR")`
   package.

# Annotation

### Updates {-}

To suggest updates and/or new mass spectrometry and/or proteomics
data, please post your suggestions/request on the
[Bioconductor support site](https://support.bioconductor.org/) or open
a
[github issue](https://github.com/lgatto/ProteomicsAnnotationHubData/issues).
Contributions can also be made to the
[package github repository](https://github.com/lgatto/ProteomicsAnnotationHubData).

### Tags {-}

A list of suggested tags is shown below. These suggestions will be
updated and completed over time.

```{r tags, echo=FALSE}
ProteomicsAnnotationHubData:::ProteomicsAnnotationHubDataTags
```

### Providers  {-}

A list of predefined/tested providers.

```{r providers, echo=FALSE}
ProteomicsAnnotationHubData:::ProteomicsAnnotationHubDataProviders
```

### Source types {-}

|---------------|---------|---------|-----------|--------------|--------|
| SourceType    |  mzML   |   mzTab |   mzid    |      FASTA   | MSnSet |
| DispatchClass | mzRpwiz |  MSnSet |  mzRident |  AAStringSet | MSnSet |
| RDataClass    | mzRpwiz |  MSnSet |  mzRident |  AAStringSet | MSnSet |

### Data location and associated metadata {-}

- The `Recipe` is a short function is a function, typically named
  `NameOfDataOrigformatToFinalformat`, that converts the original data
  into on compatible with `R`/Bioconductor.

- `Location_Prefix` is either `.amazonBaseUrl`, when the file exists
  on the AH Amazon S3 instance, or `.prideBaseUrl` when on the PRIDE
  ftp server.

- `SourceURL` is the location of the original file. This is generally
  the third-party server, but not necessarily.

- `RDataPath` is the path and filename of the file to be read into `R`
  and provided to the user. This field does not contain the server
  address `.prideBaseUrl` or `.amazonBaseUrl`.

- The metadata list, used to create the `AnnotationHubResources` also
  uses a `SourceBaseUrl`, which is the full url minus file name (that
  is in `File`) of the original file. Used to construct `SourceUrl`.

For files that need to be downloaded from a third-party provider such
as the PRIDE server and then pre-processed, the pre-processed product
being strored to Amazon s3 machine and the user directly gets the
object from Amazon S3 instance:

- the `Recipe` argument **should not be** `NA`.
- the `Location_Prefix` **should be** the `.amazonBaseUrl`.
- the `RDataPath` should correspond to the directory structure after
  `.amazonBaseUrl` on the Amazon s3 instance. Typically, the directory
  structure on the Amazon S3 instance mimics the directory structure
  on the original server.
- the `SourceUrl` should be the actual url on third-party server.


For files that are downloaded from the third-party resource and loaded
directly into `R` without any pre-processing:

- the `Recipe` argument **should be** `NA`.
- the `Location_Prefix` should be the `.prideBaseUrl`.
- the `RDdataPath` should be `sub(.prideBaseUrl, "", SourceUrl)`
- the `SourceUrl` should be the actual url on third-party server.

### Location prefix {-}

- `.prideBaseUrl` for files on the PRIDE ftp server.
- `.amazonBaseUrl` for files on `AnnotationHub` S3 instance.

### Preparer functions {-}

Preparer functions and recipes are only required if the `rda` file is
prepared on the `AnnotationHub` Amazon S3 instance.
